### PR TITLE
Issue 64

### PR DESCRIPTION
**Pull request by naomijub**
- [x] Fixed error. handling when entity functions do not return 2XX.
- [x] Upgrade `edn-rs`
- [x] Removes unnecessary check if id starts with `":"`
- [ ] Fix error handling for status code != 2xx in other functions

REMOVED:
```rust
if !crux_id.starts_with(":") {
     return Ok(edn!({:status ":bad-request", :message "ID required", :code 400}));
}
```

### CODE REVIEW

**Comment by naomijub** [comment]
https://github.com/naomijub/transistor/issues/64

File: src/http.rs [hunk 1]
```diff
@@ -86,11 +83,22 @@ impl HttpClient {
...
-            .text()?;
-
-        let edn_resp = Edn::from_str(&resp.replace("#inst", ""));
-        edn_resp.or(Ok(edn!({:status ":internal-server-error", :code 500})))
+            .send()?;
+
+        if resp.status().as_u16() < 300 {
+            let resp_body = resp.text()?;
+            let edn_resp = Edn::from_str(&resp_body.replace("#inst", ""));
+            edn_resp.or(Err(CruxError::BadRequestError(format!(
```
Comment (otaviopace) [inline src/http.rs:91]: We can use `or_else` here to avoid an unnecessary `CruxError` creation.

File: src/http.rs [hunk 2]
```diff
@@ -117,30 +122,51 @@ impl HttpClient {
...
```
Comment (otaviopace) [inline src/http.rs:130]: Here too (`or_else`)

File: src/http.rs [hunk 3]
```diff
@@ -117,30 +122,51 @@ impl HttpClient {
...
             .post(&format!("{}/entity-tx", self.uri))
             .headers(self.headers.clone())
             .body(s)
-            .send()?
-            .text()?;
-
-        EntityTxResponse::from_str(&resp.replace("#inst", ""))
+            .send()?;
+
+        if resp.status().as_u16() < 300 {
```
Comment (otaviopace) [inline src/http.rs:160]: Shoudn't we create a function to aggregate this `if`/`else` logic? :thinking: 
It seems the same every time.

File: src/http.rs [hunk 4]
```diff
@@ -86,11 +83,22 @@ impl HttpClient {
...
-        let edn_resp = Edn::from_str(&resp.replace("#inst", ""));
-        edn_resp.or(Ok(edn!({:status ":internal-server-error", :code 500})))
+            .send()?;
+
+        if resp.status().as_u16() < 300 {
+            let resp_body = resp.text()?;
+            let edn_resp = Edn::from_str(&resp_body.replace("#inst", ""));
+            edn_resp.or(Err(CruxError::BadRequestError(format!(
+                "entity responded with {} for id \"{}\" ",
+                500, crux_id
```
Comment (otaviopace) [inline src/http.rs:93]: Also, why is this a 500? I didn't get it.

Comment (naomijub) [reply src/http.rs:93]: This is legacy and just means that the process failed. not going to change this now

File: src/http.rs (hunk 1 above)
Comment (naomijub) [reply src/http.rs:91]: This is an Error situation not an OK with bad content. Maybe in a next PR create a new error.

File: src/http.rs (hunk 2 above)
Comment (naomijub) [reply src/http.rs:130]: same as above

File: src/http.rs (hunk 3 above)
Comment (naomijub) [reply src/http.rs:160]: There are 4 `if` that differ in 2 different responses, also it is not easy to deal with all the borrowing needed here. I prefer to leave the ifs, but you are welcome to try to extract it kkkkk

File: src/http.rs (hunk 4 above)
Comment (naomijub) [reply src/http.rs:93]: Also it would be super weird to return an `Err` containing a status code lower than 300

File: src/http.rs (hunk 1 above)
Comment (naomijub) [reply src/http.rs:91]: This may be an alternative CruxError https://github.com/naomijub/transistor/pull/67/files

Comment (otaviopace) [reply src/http.rs:91]: I know, is only that using `or` creates the value eagerly and `or_else` only creates if its necessary

**Review by otaviopace — APPROVED** [review]

**Pull request merged**